rllib/core/rl_trainer/trainer_runner.py

```diff
@@ -42,80 +42,84 @@ def __init__(
-                "num_gpus are specified it must be equal to num_gpus"
+
+        if num_gpus > 0:
+            scaling_config = ScalingConfig(
+                num_workers=num_gpus,
+                use_gpu=(not use_fake_gpus),
             )
 
-        elif num_gpus and not num_workers:
-            num_workers = num_gpus
+            if issubclass(trainer_class, TorchRLTrainer):
+                from ray.train.torch import TorchConfig
+
+                backend_config = TorchConfig()
+            elif issubclass(trainer_class, TfRLTrainer):
+                from ray.train.tensorflow import TensorflowConfig
 
-        elif not num_gpus and not num_workers:
-            num_workers = 1
+                backend_config = TensorflowConfig()
+            else:
+                raise ValueError("framework must be either torch or tf")
+
+            self.backend_executor = BackendExecutor(
+                backend_config=backend_config,
+                num_workers=scaling_config.num_workers,
+                num_cpus_per_worker=scaling_config.num_cpus_per_worker,
+                num_gpus_per_worker=scaling_config.num_gpus_per_worker,
+                max_retries=0,
+            )
 
-        return {"num_workers": num_workers, "use_gpu": bool(num_gpus)}
+            # TODO: let's not pass this into the config which will cause
+            # information leakage into the SARLTrainer about other workers.
+            trainer_config["distributed"] = self._distributed = True
+            self.backend_executor.start(
+                train_cls=trainer_class, train_cls_kwargs=trainer_config
+            )
+            self._workers = [
+                w.actor for w in self.backend_executor.worker_group.workers
+            ]
+
+            ray.get([w.build.remote() for w in self._workers])
+
+        else:
+            trainer_config["distributed"] = self._distributed = False
+            self._trainer = trainer_class(**trainer_config)
+            self._trainer.build()
 
-    def update(self, batch: MultiAgentBatch = None, **kwargs):
+    def update(self, batch: MultiAgentBatch) -> List[Mapping[str, Any]]:
+        """Do a gradient based update to the RLTrainer(s) maintained by this TrainerRunner.
+
+        Args:
+            batch: The data to use for the update.
+
+        Returns:
+            A list of dictionaries of results from the updates from the RLTrainer(s)
         """
-        Example:
-            >>> trainer_runner.update(batch) # updates the gradient
+        if self._distributed:
+            return self._distributed_update(batch)
+        else:
+            return [self._trainer.update(batch)]
+
+    def _distributed_update(self, batch: MultiAgentBatch) -> List[Mapping[str, Any]]:
+        """Do a gradient based update to the RLTrainers using DDP training.
+
+        Note: this function is used if the num_gpus this TrainerRunner is configured
+            with is > 0 and fake_gpus == False.
```

Comment:
It will also run when fake_gpus = True, right?

Reply:
as long as num_gpus > 0

Reply:
yep I've upd the docstring